src/effects/VST3/VST3Effect.cpp

```diff
@@ -339,32 +339,129 @@ bool VST3Effect::SupportsAutomation() const
    return false;
 }
 
-bool VST3Effect::SaveSettings(
-   const EffectSettings &, CommandParameters & parms) const
+struct VST3Effect::ParameterInfo
 {
-   if(mEditController == nullptr)
+   Steinberg::Vst::ParameterInfo mVstParamInfo;
+};
+
+
+bool VST3Effect::ForEachParameter(ParameterVisitor visitor) const
```

Comment:
This function could also be used to simplify `SupportsAutomation` and `LoadFactoryDefaults`

As for `LoadSettings`, change it to use `ForEachParameter`, and not iterating over `parms`.  Instead call `parms.Read` as in `AudioUnitEffect::LoadSettings`

Reply:
I am not sure about using it for `SupportsAutomation`, because `ForEachParameter` will always stop at the first parameter where the visitor fails, and return true - while `SupportsAutomation` stops at the first parameter which can be automated, returning true, but if none of the parameters can be automated, it will return false.

I think I'd need a different version of `ForEachParameter`, which will return true at the first successful visit and false if no successful visits happened (it could be named something like VerifyAtLeastOneParameter)

Reply:
As for `LoadSettings`, I need to have a talk about that first

Reply:
I prepared a version of `LoadSettings` that uses `ForEachParameter`, but it is deactivated (under #if 0) in favor of the existing version - this because I can not find a way to reach this method (I tried using macros, as hinted in our conversation, but the method called when playing back the macro will not take the else branch that calls LoadSettings). When I can rebase on Vitaly's recent work for stacks, I will be able to reach this method and test if the new version works.